### PR TITLE
fix the global partitioner bug

### PR DESCRIPTION
**Pull request by lanluo-nvidia**
https://github.com/pytorch/TensorRT/issues/3157

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 1]
```diff
@@ -408,6 +407,9 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
     # Generate the corresponding TRT Module for those
     for name, _ in partitioned_module.named_children():
         submodule = getattr(partitioned_module, name)
+        # filter on the GraphModule
+        if not isinstance(submodule, torch.fx.graph_module.GraphModule):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:411]: What is the logic behind this?

File: py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py [hunk 2]
```diff
@@ -228,8 +228,22 @@ def partition(
...
     # Then, fuse partitions and display overview of supported/unsupported operators
     partitions = partitioner.propose_partitions()
-    fused_graph = partitioner.fuse_partitions(partitions)
-
+    # TODO: confirm with Naren whether this change is required or not
+    # tested both with and without this change, it both works
+    # the only difference is the graph node name, an example is as below:
+    # graph():
+    # %x : [num_users=1] = placeholder[target=x]
+    # %_run_on_acc_0 : [num_users=1] = call_module[target=_run_on_acc_0](args = (%x,), kwargs = {})
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py:236]: I think we condition on _run_on_acc so the name is important @peri044 to confirm

**Review by narendasan — COMMENTED** [review]
Do we have test cases for the global partitioner?

File: py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py [hunk 3]
```diff
@@ -228,8 +228,22 @@ def partition(
     # Determine partitions based on user specifications and operator support
     # Then, fuse partitions and display overview of supported/unsupported operators
     partitions = partitioner.propose_partitions()
```
Comment (lanluo-nvidia) [inline py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py:230]: @narendasan @dheerajperi 
There is an upstream change on CapabilityBasedPartitioner from pytorch:
https://github.com/pytorch/pytorch/commit/62403b57b9810cf47955a101a1d642e244722358
which added the prefix option, I am not sure whether we need to add it.
current: (if prefix is not specified, it uses 'fused_' by default)
`fused_graph = partitioner.fuse_partitions(partitions)`
should I to change to:
`fused_graph = partitioner.fuse_partitions(partitions, prefix="_run_on_acc_")`
I have tested both with change or without change, it both works, the only difference is the graph node name:
 ```
 graph():
   %x : [num_users=1] = placeholder[target=x]
   %_run_on_acc_0 : [num_users=1] = call_module[target=_run_on_acc_0](args = (%x,), kwargs = {})
   return (_run_on_acc_0,)
```

    or

   ```
graph():
   %x : [num_users=1] = placeholder[target=x]
   %fused_0 : [num_users=1] = call_module[target=fused_0](args = (%x,), kwargs = {})
   return (fused_0,)
```

File: py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py (hunk 2 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py:236]: Yes, we condition on that at a couple of places in the code
1) https://github.com/search?q=repo%3Apytorch%2FTensorRT+_run_on_acc+path%3A%2F%5Etests%5C%2Fpy%5C%2Fdynamo%5C%2F%2F&type=code
2) https://github.com/search?q=repo%3Apytorch%2FTensorRT+_run_on_acc+path%3A%2F%5Epy%5C%2Ftorch_tensorrt%5C%2Fdynamo%5C%2F%2F&type=code
Using same prefix for both the partitioners would be better.

**Review by peri044 — COMMENTED** [review]
seems fine to me (after the prefix discussion). I have same questions as Naren

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 1 above)
Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/_compiler.py:411]: partitioned_module has named_children() submodule which is not torch.fx.graph_module.GraphModule type, we should only filter on the torch.fx.graph_module.GraphModule to be convertable to TRT Module.

File: tests/py/dynamo/partitioning/test_global_partitioning.py [hunk 4]
```diff
@@ -1,14 +1,52 @@
...
+        with torch.no_grad():
+            inputs = torch.randn(
+                batch_size, 3, tile_size, tile_size, device="cuda", dtype=torch.float16
+            )
+            try:
+                torch_tensorrt.compile(
+                    mod,
+                    ir="dynamo",
+                    inputs=[inputs],
+                    enabled_precisions={torch.float16},
```
Comment (narendasan) [inline tests/py/dynamo/partitioning/test_global_partitioning.py:44]: You should probably force fallback an op or two to make sure there are at least a few subgraphs

Comment (lanluo-nvidia) [reply tests/py/dynamo/partitioning/test_global_partitioning.py]: added, please review again.
